### PR TITLE
IR fix and example

### PR DESCRIPTION
**Pull request by fritzgoebel**
This PR introduces the following fix to IR:
When using the advanced apply for the inner solver, the initial guess will be `x`. This way, the initial residual of the inner solver is `r - Ax` which will generally not be small but close to `Ax` as we try to get `r` to be zero. This causes very poor to no convergence when e.g. using CG with tolerance 1e-2 as inner solver. By kind of manually doing the advanced apply, we can use 'r' as initial guess, which gives much better results.

This PR also introduces a small example on iterative refinement.

### CODE REVIEW

File: examples/iterative-refinement/data/A.mtx [hunk 1]
```diff
@@ -0,0 +1,114 @@
+%%MatrixMarket matrix coordinate integer symmetric
```
Comment (pratikvn) [inline examples/iterative-refinement/data/A.mtx:1]: Maybe we can move this from all the specific examples directories to a common directory. Now we just duplicate this matrix everywhere.

File: examples/iterative-refinement/iterative-refinement.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    auto A = share(gko::read<mtx>(std::ifstream("data/A.mtx"), exec));
+    // Create RHS and initial guess as 1
+    gko::size_type size = A->get_size()[0];
+    auto host_x = gko::matrix::Dense<ValueType>::create(exec->get_master(),
+                                                        gko::dim<2>(size, 1));
+    for (auto i = 0; i < size; i++) {
+        host_x->at(i, 0) = 1.;
+    }
+    auto x = gko::matrix::Dense<>::create(exec);
+    auto b = gko::matrix::Dense<>::create(exec);
```
Comment (pratikvn) [inline examples/iterative-refinement/iterative-refinement.cpp:83]:
```suggestion
    auto x = gko::matrix::Dense<ValueType>::create(exec);
    auto b = gko::matrix::Dense<ValueType>::create(exec);
```

File: examples/iterative-refinement/iterative-refinement.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    b->compute_norm2(lend(initres));
+
+    // copy b again
+    b->copy_from(host_x.get());
+
+    auto iter_stop =
+        gko::stop::Iteration::build().with_max_iters(10000u).on(exec);
+    auto tol_stop = gko::stop::ResidualNormReduction<ValueType>::build()
+                        .with_reduction_factor(static_cast<ValueType>(1e-12))
+                        .on(exec);
```
Comment (pratikvn) [inline examples/iterative-refinement/iterative-refinement.cpp:101]: This allows us to support complex types and also to tweak the iteration count better. I am not sure about the max iter count, I would set it to the number of rows of the matrix as that is usually a good estimate for Krylov solvers. But for IR maybe it is different.

```suggestion
   gko::size_type max_iters= 10000u; 
    gko::remove_complex<ValueType> outer_reduction_factor = 1e-12;
    auto iter_stop =
        gko::stop::Iteration::build().with_max_iters(max_iters).on(exec);
    auto tol_stop = gko::stop::ResidualNormReduction<ValueType>::build()
                        .with_reduction_factor(outer_reduction_factor)
                        .on(exec);
```

File: examples/iterative-refinement/iterative-refinement.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    tol_stop->add_logger(logger);
+
+    // Create solver factory
+    auto solver_gen =
+        ir::build()
+            .with_solver(
+                cg::build()
+                    .with_criteria(
+                        gko::stop::ResidualNormReduction<ValueType>::build()
+                            .with_reduction_factor(static_cast<ValueType>(1e-2))
```
Comment (pratikvn) [inline examples/iterative-refinement/iterative-refinement.cpp:115]:
```suggestion
    // Create solver factory
    gko::remove_complex<ValueType> inner_reduction_factor = 1e-2;
    auto solver_gen =
        ir::build()
            .with_solver(
                cg::build()
                    .with_criteria(
                        gko::stop::ResidualNormReduction<ValueType>::build()
                            .with_reduction_factor(inner_reduction_factor)
```

File: examples/iterative-refinement/iterative-refinement.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    // Calculate initial residual by overwriting b
+    auto one = gko::initialize<vec>({1.0}, exec);
+    auto neg_one = gko::initialize<vec>({-1.0}, exec);
+    auto initres = gko::initialize<vec>({0.0}, exec);
+    A->apply(lend(one), lend(x), lend(neg_one), lend(b));
+    b->compute_norm2(lend(initres));
+
+    // copy b again
+    b->copy_from(host_x.get());
+   gko::size_type max_iters= 10000u;
```
Comment (pratikvn) [inline examples/iterative-refinement/iterative-refinement.cpp:96]:
```suggestion
    gko::size_type max_iters= 10000u; 
```
Sorry, my fault.

**Comment by hartwiganzt** [comment]
@fritzgoebel I think this is ready to merge...

**Comment by upsj** [comment]
@fritzgoebel Maybe before merging, can you do an interactive rebase onto develop and clean up the history a bit? Especially the "Update examples/*" files

```bash
git fetch   # copy changes from remote
git rebase -i origin/develop
```

**Pull request merged**